src/groups/mqb/mqbblp/mqbblp_cluster.cpp

```diff
@@ -3674,5 +3715,99 @@ void Cluster::processResponse(const bmqp_ctrlmsg::ControlMessage& response)
         this);
 }
 
+void Cluster::getPrimaryNodes(bsl::vector<mqbnet::ClusterNode*>* outNodes,
+                              bool* outIsSelfPrimary) const
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(outNodes);
+    BSLS_ASSERT_SAFE(outIsSelfPrimary);
+    BSLS_ASSERT_SAFE(dispatcher()->inDispatcherThread(this));
+
+    const mqbnet::Cluster::NodesList&         nodes = netCluster().nodes();
+    const mqbc::ClusterState::PartitionsInfo& partitionsInfo =
+        d_state.partitions();
+
+    *outIsSelfPrimary = false;
+    outNodes->clear();
+
+    for (mqbc::ClusterState::PartitionsInfo::const_iterator pit =
+             partitionsInfo.begin();
+         pit != partitionsInfo.end();
+         pit++) {
+        if (pit->primaryStatus() !=
+            // TODO: Handle this case (will want to buffer)
+            bmqp_ctrlmsg::PrimaryStatus::Value::E_ACTIVE) {
+            BALL_LOG_WARN << "While collecting primary nodes: Partition id "
+                          << "primary for partition " << pit->partitionId()
+                          << " is not active";
+            continue;
+        }
+
+        mqbnet::ClusterNode* primary = pit->primaryNode();
+
+        if (primary) {
+            // Don't add duplicate
+            if (bsl::find(outNodes->begin(), outNodes->end(), primary) !=
+                outNodes->end()) {
+                continue;
+            }
+            // Check for self
+            if (d_state.isSelfActivePrimary(pit->partitionId())) {
+                *outIsSelfPrimary = true;
+                continue;
+            }
+            outNodes->push_back(primary);
+        }
+        else {
+            // TODO: handle this case
```

Comment:
Right now the response is empty in this case, the idea is that this would be fixed when I implement the buffer and retry (or error if retry fails). But that's not implemented in this PR, so I can go ahead and implement just an error message in this case for now.